src/main/java/com/puppycrawl/tools/checkstyle/CheckCloneService.java

```diff
@@ -0,0 +1,156 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.api.Configurable;
+import com.puppycrawl.tools.checkstyle.api.Configuration;
+import com.puppycrawl.tools.checkstyle.api.Contextualizable;
+import com.puppycrawl.tools.checkstyle.api.FileSetCheck;
+
+/**
+ * A service for clonning check instances.
+ * @author Andrew Kuchev.
+ */
+public final class CheckCloneService {
+    /** Logger for CheckCloneService. */
+    private static final Log LOG = LogFactory.getLog(CheckCloneService.class);
+
+    /** A log pattern for case when a check does not implement appropriate interface. */
+    private static final String CHECK_MUST_IMPLEMENT_INTERFACE_MSG =
+            "Check %s should implement either OneCheckInstancePerThread "
+                    + "or OneCheckInstancePerApplication%n";
+
+    /** A private constructor is required for utility class. */
+    private CheckCloneService() {
+    }
+
+    /**
+     * Clones the given check.
+     * @param check A check to be cloned.
+     * @return Cloned check.
+     * @noinspection InstanceofChain
+     */
+    // we cannot move this logic to the AbstractCheck subclasses,
+    // because they should not know about PackageObjectFactory
+    // and other configuration details
+    public static AbstractCheck cloneCheck(AbstractCheck check) {
+        final AbstractCheck clone;
+        if (check instanceof OneCheckInstancePerApplication) {
+            clone = check;
+        }
+        else if (check instanceof OneCheckInstancePerThread) {
+            final ClassLoader classLoader = CheckCloneService.class.getClassLoader();
+            final ModuleFactory factory = new PackageObjectFactory(
+                    Checker.class.getPackage().getName(), classLoader);
+            clone = doCloneCheck(check, factory);
+            clone.init();
+        }
+        else {
+            if (LOG.isDebugEnabled()) {
+                final String checkClassName = check.getClass().getSimpleName();
+                final String message = String.format(Locale.ROOT,
+                        CHECK_MUST_IMPLEMENT_INTERFACE_MSG, checkClassName);
+                LOG.debug(message);
+            }
+            clone = check;
```

Comment:
Shouldn't the default clone be the strongest cloning, which is `OneCheckInstancePerThread` and not `OneCheckInstancePerApplication`? This would cause the least amount of trobules for 3rd party checks who have not upgraded.